src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs

```diff
@@ -114,59 +148,131 @@ bool IsValidTriggerCharacterForDocument(Document document, char triggerCharacter
                 {
-                    var vsCompletionItem = CreateCompletionItem<LSP.VSCompletionItem>(request, item, resultId, completionTrigger, commitCharacterRulesCache);
+                    var vsCompletionItem = await CreateCompletionItemAsync<LSP.VSCompletionItem>(
+                        request, document, item, resultId, completionTrigger, commitCharacterRulesCache,
+                        completionService, clientName, returnTextEdits, documentText, defaultSpan,
+                        defaultRange, cancellationToken).ConfigureAwait(false);
                     vsCompletionItem.Icon = new ImageElement(item.Tags.GetFirstGlyph().GetImageId());
                     return vsCompletionItem;
                 }
                 else
                 {
-                    var roslynCompletionItem = CreateCompletionItem<LSP.CompletionItem>(request, item, resultId, completionTrigger, commitCharacterRulesCache);
+                    var roslynCompletionItem = await CreateCompletionItemAsync<LSP.CompletionItem>(
+                        request, document, item, resultId, completionTrigger, commitCharacterRulesCache,
+                        completionService, clientName, returnTextEdits, documentText, defaultSpan,
+                        defaultRange, cancellationToken).ConfigureAwait(false);
                     return roslynCompletionItem;
                 }
             }
 
-            static TCompletionItem CreateCompletionItem<TCompletionItem>(
+            static async Task<TCompletionItem> CreateCompletionItemAsync<TCompletionItem>(
                 LSP.CompletionParams request,
+                Document document,
                 CompletionItem item,
                 long? resultId,
                 CompletionTrigger completionTrigger,
-                Dictionary<ImmutableArray<CharacterSetModificationRule>, ImmutableArray<string>> commitCharacterRulesCache) where TCompletionItem : LSP.CompletionItem, new()
+                Dictionary<ImmutableArray<CharacterSetModificationRule>, ImmutableArray<string>> commitCharacterRulesCache,
+                CompletionService completionService,
+                string? clientName,
+                bool returnTextEdits,
+                SourceText? documentText,
+                TextSpan? defaultSpan,
+                LSP.Range? defaultRange,
+                CancellationToken cancellationToken) where TCompletionItem : LSP.CompletionItem, new()
             {
                 var completeDisplayText = item.DisplayTextPrefix + item.DisplayText + item.DisplayTextSuffix;
+
+                // The TextEdits for override and partial method completions are provided in the resolve handler.
+                // We do not provide them in this handler.
+                item.Properties.TryGetValue("Modifiers", out var modifier);
+                var isOverrideOrPartialMethodCompletion = modifier != null && (modifier.Contains("Override") || modifier.Contains("Partial"));
```

Comment:
this feels like very brittle and fragile coupling.  could we instead have some new interface that these providers can implement to say "do X with me" or "don't do Y with me".  Similar to how we have LSPCompletionProvider to provide the trigger characters we need for LSP.

Reply:
it seems a little harder here since the completion service just returns the completion item, we don't have access to the specific provider that creates the completion item, it's abstracted behind the completion service.  Maybe we could have the provider add a well known tag / property marking the items it creates as needing complex text edit resolution?

Reply:
It looks like many other places in existing Roslyn code rely on looking up properties the same way we're doing here. For example, [here](http://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp.Features/Completion/CompletionProviders/CrefCompletionProvider.cs,423), [here](http://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp.Features/Completion/CompletionProviders/PartialTypeCompletionProvider.cs,90), [here](http://sourceroslyn.io/#Microsoft.CodeAnalysis.EditorFeatures/Implementation/IntelliSense/AsyncCompletion/CompletionSource.cs,325), and [here](http://sourceroslyn.io/#Microsoft.CodeAnalysis.EditorFeatures/Implementation/IntelliSense/AsyncCompletion/CompletionSource.cs,468).

Because this behavior is already prevalent throughout the repo, IMO finding a solution to this seems out of scope for this PR. I opened up #51396 to track this.

Reply:
hrmm...  ok.  we can have an issue on this.  however, we should doc this as a HACK, and link to the issue please.  thanks :)